src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,151 @@
+
+    public Shooter() {
+        // Configure the motors
+        main.setInverted(Ports.Shooter.MAIN_INVERTED);
+        aux.setInverted(Ports.Shooter.AUX_INVERTED);
+        main.setSensorPhase(Ports.Shooter.IS_SENSOR_INVERTED);
+
+        main.setNeutralMode(NeutralMode.Coast);
+        aux.setNeutralMode(NeutralMode.Coast);
+
+        main.configForwardLimitSwitchSource(LimitSwitchSource.Deactivated, LimitSwitchNormal.Disabled);
+        main.configReverseLimitSwitchSource(LimitSwitchSource.Deactivated, LimitSwitchNormal.Disabled);
+
+        main.enableVoltageCompensation(true);
+        aux.enableVoltageCompensation(true);
+
+        main.configVoltageCompSaturation(Constants.NOMINAL_VOLTAGE);
+        aux.configVoltageCompSaturation(Constants.NOMINAL_VOLTAGE);
+
+        aux.follow(main);
+
+        this.stateSpacePredictor = constructLinearSystem();
+        this.velocityEstimator = new MovingAverage(readCSV());
+    }
+
+    private LinearSystemLoop<N1, N1, N1> constructLinearSystem() {
+        // https://file.tavsys.net/control/controls-engineering-in-frc.pdf Page 76
+        Vector<N1> A = VecBuilder.fill(-Math.pow(GEAR_RATIO, 2) * Kt / (Kv * OMEGA * J)); //Change the amount of cells and rows
+        Vector<N1> B = VecBuilder.fill(GEAR_RATIO * Kt / (OMEGA * J));
+        LinearSystem<N1, N1, N1> stateSpace = new LinearSystem<>(A, B, Matrix.eye(Nat.N1()), new Matrix<>(Nat.N1(), Nat.N1()));
+        KalmanFilter<N1, N1, N1> kalman = new KalmanFilter<>(Nat.N1(), Nat.N1(), stateSpace,
+                VecBuilder.fill(MODEL_TOLERANCE),
+                VecBuilder.fill(ENCODER_TOLERANCE),
+                Constants.LOOP_PERIOD
+        );
+        LinearQuadraticRegulator<N1, N1, N1> lqr = new LinearQuadraticRegulator<>(stateSpace, VecBuilder.fill(VELOCITY_TOLERANCE),
+                VecBuilder.fill(Constants.NOMINAL_VOLTAGE),
+                Constants.LOOP_PERIOD // time between loops, DON'T CHANGE
+        );
+
+        return new LinearSystemLoop<>(stateSpace, lqr, kalman, Constants.NOMINAL_VOLTAGE, Constants.LOOP_PERIOD);
+    }
+
+    private InputStreamReader readCSV() {
+        InputStream is = getClass().getResourceAsStream(PATH_TO_CSV);
+        assert is != null : "Can't create input stream";
+
+        return new InputStreamReader(is);
+    }
+
+    /**
+     * @return the velocity of the shooter. [RPS]
+     * @see #setVelocity(double)
+     */
+    public double getVelocity() {
+        return unitModel.toVelocity(main.getSelectedSensorVelocity());
+    }
+
+    /**
+     * Set the velocity to apply by the motor.
+     *
+     * @param velocity the desired velocity at which the motor will rotate. [RPS]
+     * @see #setPower(double)
+     */
+    public void setVelocity(double velocity) {
+        stateSpacePredictor.setNextR(VecBuilder.fill(velocity)); //r = reference
+        stateSpacePredictor.correct(VecBuilder.fill(getVelocity()));
+        stateSpacePredictor.predict(Constants.LOOP_PERIOD); //every 20 ms
```

Comment:
It's a few additional lines of code, but I believe passing the actual timestamp instead of a constant will be more accurate.